Fixes #2743 Horizontal wrapper can't be created with 4.1.16

dbeltrankyl opened the pull request:
Closes #2743

<!--

Thanks for your contribution! Please:
* List any related issues with a "closes" or "addresses" tag.
* Add a helpful title & description.
* Complete the checklist.

-->

**Check List**

- [x] I have read `CONTRIBUTING.md`.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to `pyproject.toml`.
- [x] Tests are included (or explain why tests are not needed).
- [x] Changelog entry included in `CHANGELOG.md` if this is a change that can affect users.
- [x] Documentation updated.
- [x] If this is a bug fix, PR should include a link to the issue (e.g. `Closes #1234`).

dbeltrankyl commented on autosubmit/config/configcommon.py:
```diff
@@ -2871,6 +2871,6 @@ def get_current_wrapper(self, section: str) -> dict:
         """
         for wrapper in self.experiment_data.get("WRAPPERS", {}).values():
-            if section in wrapper.get("JOBS_IN_WRAPPER", ""):
+            if isinstance(wrapper, dict) and section in wrapper.get("JOBS_IN_WRAPPER", ""):
```
This is evaluating this dict:

```yaml
WRAPPERS:
 WRAPPER: # this is always a dict
    jobs_in_wrapper: sim 
 OTHER_WRAPPER: # this is always a dict
    jobs_in_wrapper: other_job
 other_parameter_like_min_wrapped: 20 
    
 ````
 Here we have:
         
> for wrapper in self.experiment_data.get("WRAPPERS", {}).values():
 
 ```yaml
WRAPPER: # this is always a dict
    jobs_in_wrapper: sim 
OTHER_WRAPPERS: # this is always a dict
    other_parameter: 20
other_parameter_like_min_wrapped: 20 
 ```
 
 >  if isinstance(wrapper, dict) and section in wrapper.get("JOBS_IN_WRAPPER", ""):
 
 1.iteration. 
 
 WRAPPER is a dict? Yes, does it include jobs_in_wrapper? Yes, check if the current job. section is inside the WRAPPER section ( to get the current_wrapper for this job.section)
 
 2. iteration:
 
 OTHER_WRAPPER is a dict? Yes, does it include jobs_in_wrapper? no
 
 
 3. iteration:
 
 Without the fix: 
  other_parameter_like_min_wrapped is a dict? No, crash. 
  
  
With the fix:
  other_parameter_like_min_wrapped is a dict? no. continue

kinow commented on test/integration/commands/run/test_run_slurm_scenarios.py:
```diff
@@ -486,3 +486,137 @@ def test_run_with_additional_files(
+            {
+                "WRAPPERS": {
+                    "WRAPPER": {
+                        "JOBS_IN_WRAPPER": "job&other",
```
👍 I remember somewhere we had an issue with & or spaces not working. Good to have more tests with these different syntax options!

kinow commented on autosubmit/autosubmit.py:
```diff
@@ -1760,13 +1759,10 @@ def inspect(expid: str, lst: str, filter_chunks: str, filter_status: str, filter
-                    wrapped_sections = list(set(wrapped_sections))
+                        if isinstance(wrapper_data, dict):
+                            jobs_in_wrapper = wrapper_data.get("JOBS_IN_WRAPPER", [])
+                            wrapped_sections.extend(jobs_in_wrapper)
```
Can´t  `jobs_in_wrapper` here be something like `sim&dn` or `sim,dn`? Asking only because the code removed appeared to be covering at least commas and spaces...

kinow replied:
I also think this is not being tested

<img width="1409" height="777" alt="Image" src="https://github.com/user-attachments/assets/2198c235-0fa4-4fa8-8ca1-77be73c4ceb7" />

dbeltrankyl replied:
No, here it will be normalized already in the as_conf.reload() -> normalize_data -> normalize_wrappers

Oh, I see why it is not covered, it is inside the --quick option, I can add it easily

dbeltrankyl commented on test/unit/config/test_normalize_variables.py:
```diff
@@ -420,3 +422,74 @@ def test_normalize_variables(autosubmit_config, data, expected_data, must_exists
+
+    as_conf = autosubmit_config(expid='t000', experiment_data=input_data)
+    with pytest.raises(AutosubmitCritical):
+        as_conf._normalize_wrappers_section(input_data)
```
I was only checking if we received a string or a list.  You're right that [1,2,3,4..] would be valid as well. 

These are the cases covered: 

```yaml
jobs_in_wrapper: 
 - bla
  -bla 
  
jobs_in_wrapper: bla

jobs_in_wrapper:  bla, bla

jobs_in_wrapper:  ["bla", "bla"] # unlikely 

jobs_in_wrapper:  bla&bla

```
I wanted to check the cases, which are the only ones I saw... maybe we can improve them in another PR?

dbeltrankyl commented on autosubmit/config/configcommon.py:
```diff
@@ -553,11 +553,24 @@ def _normalize_wrappers_section(data_fixed: dict) -> None:
+                        jobs_in_wrapper = jobs_in_wrapper.split("&")
+                    else:
+                        jobs_in_wrapper = jobs_in_wrapper.split()
+                data_fixed["WRAPPERS"][wrapper]["JOBS_IN_WRAPPER"] = [job.upper().strip(" ,") for job in jobs_in_wrapper]
```
Yup, all resolved

dbeltrankyl commented on autosubmit/config/configcommon.py:
```diff
@@ -553,11 +553,24 @@ def _normalize_wrappers_section(data_fixed: dict) -> None:
+                    if "," in jobs_in_wrapper:
+                        jobs_in_wrapper = jobs_in_wrapper.split(",")
+                    elif "&" in jobs_in_wrapper:
+                        jobs_in_wrapper = jobs_in_wrapper.split("&")
```
Hello @kinow 

Solved! added "_" in my tests 

This issue wasn't related to my changes at all. I'm glad that you spotted it! Because we would have issues with OPA_blablabla


All that the lambda function was doing was to grab the job.chunk number from the job.name if it is not None. So I simplified it to look for the job.chunk.

dbeltrankyl replied:
<img width="867" height="244" alt="image" src="https://github.com/user-attachments/assets/4b4088ed-1b4e-4ecb-ae70-1fab1731b17e" />

Tested and with splits it looks sorted well 👍

kinow commented on autosubmit/job/job_packages.py:
```diff
@@ -52,16 +52,9 @@ def wrapper(*args, **kwargs):
-    jobs_in_wrapper = [job.strip(" ,") for job in jobs_in_wrapper]
-    return "_".join(jobs_in_wrapper)
+def jobs_in_wrapper_str(as_conf: 'AutosubmitConfig', current_wrapper: str) -> list[str]:
+    " transform to string with _ separator the jobs in the wrapper "
```
This is still a string,

```suggestion
    """Transform to string with _ separator the jobs in the wrapper."""
```

kinow commented on autosubmit/config/configcommon.py:
```diff
@@ -550,14 +550,42 @@ def _normalize_default_section(self, data_fixed: dict) -> None:
+                sanitazed_jobs_in_wrapper = [job.upper().strip(" ,") for job in jobs_in_wrapper]
+                if jobs_in_wrapper is not None:
+                    for element in sanitazed_jobs_in_wrapper:
+                        if not isinstance(element, str):
```
Is that possible? I think because `sanitized_jobs_in_wrapper` is created with `[str.upper().strip() ...]`, that means the type is `list[str]`.

The elements can be only strings? No `None`'s, nor other objects, I think? I think this whole `_normalize_wrappers_section` function can be unit tested, which is faster. And I **think** trying to reach this `if` wouldn't be possible, confirming it's safe to remove...

dbeltrankyl commented:
Hello @kinow 

Thanks for the feedback

I'll wrap up the answers with a general comment: 

Added:

```python
 @staticmethod
    def _normalize_jobs_in_wrapper(
        wrapper: str,
        wrapper_data: dict[str, Any],
        job_sections: Iterable[str],
        raise_exception: bool,
    ) -> None:
        """Normalize the JOBS_IN_WRAPPER field for a wrapper.

        Ensure the wrapper includes a non-empty list of valid job section names,
        optionally raising if the field is missing or invalid.
        :param wrapper: The name of the wrapper being normalized.
        :param wrapper_data: The data dictionary for the specific wrapper.
        :param job_sections: The valid job section names to check against.
        :param raise_exception: If true, raise exception on errors. It is only True after all data is loaded.
        :return: None
        """
        jobs_in_wrapper = wrapper_data.get("JOBS_IN_WRAPPER", None)

        if raise_exception and not jobs_in_wrapper:
            raise AutosubmitCritical(f"JOBS_IN_WRAPPER in WRAPPERS.{wrapper} is missing or empty. This is a mandatory parameter.", 7014)

        elif raise_exception and not isinstance(jobs_in_wrapper, list) and not isinstance(jobs_in_wrapper, str):
            raise AutosubmitCritical(
                f"JOBS_IN_WRAPPER in WRAPPERS.{wrapper} must be a list or a string",
                7014
            )

        if isinstance(jobs_in_wrapper, str):
            # if it is a list in string format (due to "%" in the string).
            if "[" in jobs_in_wrapper:
                jobs_in_wrapper = jobs_in_wrapper.strip("[]").replace("'", "").replace(" ", "").replace(",", " ")
            if "," in jobs_in_wrapper:
                jobs_in_wrapper = jobs_in_wrapper.split(",")
            elif "&" in jobs_in_wrapper:
                jobs_in_wrapper = jobs_in_wrapper.split("&")
            else:
                jobs_in_wrapper = jobs_in_wrapper.split()

        sanitized_jobs_in_wrapper = []
        for job in jobs_in_wrapper:
            if isinstance(job, str):
                sanitized_jobs_in_wrapper.append(job.upper().strip())
            else:
                sanitized_jobs_in_wrapper.append(job)

        if raise_exception:
            for element in sanitized_jobs_in_wrapper:
                if not isinstance(element, str):
                    raise AutosubmitCritical(f"JOBS_IN_WRAPPER in WRAPPERS.{wrapper} must be a list of strings", 7014)
                elif len(element) == 0:
                    raise AutosubmitCritical(f"JOBS_IN_WRAPPER in WRAPPERS.{wrapper} contains empty job names ( check for double ,, or double && )", 7014)
                elif element not in job_sections:
                    raise AutosubmitCritical(f"JOBS_IN_WRAPPER in WRAPPERS.{wrapper} contains job: {element} that is not defined in JOBS section", 7014)

        wrapper_data["JOBS_IN_WRAPPER"] = sanitized_jobs_in_wrapper
 ```
 
 NOTES: 
 
 * There is now an optional "raise_exception" 
 
 Why? I noticed that the normalization is called during the unification of all files, so the wrapper could be incomplete during, but not after, all is loaded. So the raise exception code is now only called at the end.
 
 Improved the unit tests:
 
 * The tests for this are now parametrized, because I noticed that the test wasn't reaching some branches because another exception was caught earlier
 
 Integration tests:
 
  - Clean up the test, now it more simplier
 
 
 ```python
 @pytest.mark.parametrize(
    "wrappers",
    [
        pytest.param({"WRAPPER1": {"JOBS_IN_WRAPPER": {"bla": "bla"}, "TYPE": "VERTICAL"}}, id="mapping_not_allowed"),
        pytest.param({"WRAPPER2": {"JOBS_IN_WRAPPER": 12345, "TYPE": "VERTICAL"}}, id="non_iterable_value"),
        pytest.param({"WRAPPER3": {"JOBS_IN_WRAPPER": None, "TYPE": "VERTICAL"}}, id="none_value"),
        pytest.param({"WRAPPER4": {"TYPE": "VERTICAL"}}, id="missing_key"),
        pytest.param({"WRAPPER5": {"JOBS_IN_WRAPPER": ["JOB1", "NON_EXISTENT_JOB"], "TYPE": "VERTICAL"}}, id="unknown_job"),
        pytest.param({"WRAPPER6": {"JOBS_IN_WRAPPER": [], "TYPE": "VERTICAL"}}, id="empty_list"),
        pytest.param({"WRAPPER7": {"JOBS_IN_WRAPPER": ["", " "], "TYPE": "VERTICAL"}}, id="blank_entries"),
        pytest.param({"WRAPPER8": {"JOBS_IN_WRAPPER": ["JOB1", 123]}, "TYPE": "VERTICAL"}, id="non_string_job_name"),
        pytest.param({"WRAPPER9": {"JOBS_IN_WRAPPER": ["JOB1"]}}, id="type_not_found"),
    ],
)
def test_normalize_wrappers_raise_error(autosubmit_config, wrappers):
    input_data = {
        "JOBS": {"JOB1": {}, "JOB2": {}},
        "WRAPPERS": wrappers,
    }

    as_conf = autosubmit_config(expid='t000', experiment_data=input_data)
    with pytest.raises(AutosubmitCritical):
        as_conf._normalize_wrappers_section(input_data, raise_exception=True)
```

If the type is not found, it will also raise an exception now ( at the end of execution)

I've put a debug point in each if, and it's now reached. 

Ready to review agian ( if CI/CD pass)

dbeltrankyl commented on autosubmit/config/configcommon.py:
```diff
@@ -549,16 +550,77 @@ def _normalize_default_section(self, data_fixed: dict) -> None:
+        wrapper: str,
+        wrapper_data: dict[str, Any],
+        job_sections: Iterable[str],
+        raise_exception: bool,
```
Thanks @kinow !

> Or is it because load_config is called and uses this normalize... but in reality it hasn't processed all files yet? If that's the case, 

That's it

> I think it'd make more sense to perform the normalization after the files are all loaded (which would be for a future issue, I think).

The problem with not normalizing after each file is loaded is that we risk inconsistencies when unifying the configuration into a single dictionary, but this can be studied in another issue. ( In fact, at the start, we didn't have this normalization after each load, which caused issues)

load: aaaa.yml -> bbbb.yml -> unify(aaa_normalized, bbb_normalized)


Can we merge it then?
